### PR TITLE
SDT-31: Configure APIM

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/SDT-31

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

File: infrastructure/apim.tf [hunk 1]
```diff
@@ -51,3 +51,50 @@ module "api-mgmt-api-policy" {
...
 }
+
+data "azurerm_api_management" "cft-api-mgmt" {
+  name                = local.api_mgmt_name
+  resource_group_name = local.api_mgmt_resource_group
+}
+
+data "azurerm_key_vault" "key_vault" {
+  name                = "${var.product}-${var.env}"
+  resource_group_name = "${var.product}-${var.env}"
```
Comment (paulridout) [inline infrastructure/apim.tf:62]: To be consistent with the rest of the configuration add entries to the locals section for these values and then use them here, i.e. in locals:
```
key_vault_name = ${var.product}-${var.env}
key_vault_resource_group = ${var.product}-${var.env}
```
Them here:
```
name                = local.key_vault_name
resource_group_name = local.key_vault_resource_group
```

File: infrastructure/apim.tf [hunk 2]
```diff
@@ -51,3 +51,50 @@ module "api-mgmt-api-policy" {
...
+}
+
+data "azurerm_client_config" "current" {}
+
+# set key vault permissions
```
Comment (paulridout) [inline infrastructure/apim.tf:67]: Cosmetic - Capitalise first letter of first word in comment.

File: infrastructure/apim.tf [hunk 3]
```diff
@@ -51,3 +51,50 @@ module "api-mgmt-api-policy" {
...
+    "get"
+  ]
+
+  certificate_permissions = [
+    "get",
+    "list"
+  ]
+}
+
+
```
Comment (paulridout) [inline infrastructure/apim.tf:84]: Cosmetic - Remove this blank line so there's only one blank line separating the two statements.

File: infrastructure/apim.tf (hunk 1 above)
Comment (kiran-yenigala-hmcts) [reply infrastructure/apim.tf:62]: done

File: infrastructure/apim.tf (hunk 2 above)
Comment (kiran-yenigala-hmcts) [reply infrastructure/apim.tf:67]: :) done

File: infrastructure/apim.tf (hunk 3 above)
Comment (kiran-yenigala-hmcts) [reply infrastructure/apim.tf:84]: done

**Review by paulridout — DISMISSED** [review]
Changes look OK.

File: infrastructure/apim.tf [hunk 4]
```diff
@@ -1,13 +1,54 @@
...
+      subscription = "1c4f0704-a29e-403d-b719-b90c34ef14c9"
+    }
+  }
 }
 
+provider "azurerm" {
+  subscription_id            = local.cft_subscription_id[var.env].subscription
+  skip_provider_registration = "true"
+  features {}
+  alias = "acr"
```
Comment (thomast1906) [inline infrastructure/apim.tf:48]: probably rename alias to something more meaningful for this team/project

File: infrastructure/apim.tf [hunk 5]
```diff
@@ -1,13 +1,54 @@
 locals {
-  api_mgmt_name = "cft-api-mgmt-${var.env}"
+  key_vault_name = "${var.product}-${var.env}"
+  key_vault_resource_group = "${var.product}-service-${var.env}"
+  api_mgmt_name = var.env == "aat" ? "cft-api-mgmt-stg" : "cft-api-mgmt-${var.env}"
```
Comment (paulridout) [inline infrastructure/apim.tf:4]: Just to be consistent use```${var.env}``` rather than just ```var.env```

File: infrastructure/apim.tf [hunk 6]
```diff
@@ -1,13 +1,54 @@
...
   api_mgmt_resource_group = "cft-${var.env}-network-rg"
   api_mgmt_product_name = "${var.product}-${var.component}"
   api_mgmt_api_name = "${var.product}-${var.component}-api"
   api_base_path = var.product
```
Comment (paulridout) [inline infrastructure/apim.tf:8]: Just to be consistent use ```${var.product}``` instead of ```var.product```

File: infrastructure/apim.tf [hunk 7]
```diff
@@ -1,13 +1,54 @@
...
+    ptl = {
+      subscription = "1baf5470-1c3e-40d3-a6f7-74bfbce4b348"
+    }
+    prod = {
+      subscription = "8999dec3-0104-4a27-94ee-6588559729d1"
+    }
+    demo = {
+      subscription = "1c4f0704-a29e-403d-b719-b90c34ef14c9"
+    }
+  }
```
Comment (paulridout) [inline infrastructure/apim.tf:41]: These may not be needed - see https://github.com/hmcts/civil-sdt-gateway/blob/master/infrastructure/main.tf#L1

File: infrastructure/apim.tf [hunk 8]
```diff
@@ -1,13 +1,54 @@
...
 }
 
+provider "azurerm" {
+  subscription_id            = local.cft_subscription_id[var.env].subscription
+  skip_provider_registration = "true"
+  features {}
+  alias = "sdt"
+}
```
Comment (paulridout) [inline infrastructure/apim.tf:49]: There main.tf file (https://github.com/hmcts/civil-sdt-gateway/blob/master/infrastructure/main.tf) is currently being used to declare providers.  Should this declaration be moved there?

Change the alias to make it clear this is for accessing key vaults

File: infrastructure/apim.tf (hunk 4 above)
Comment (kiran-yenigala-hmcts) [reply infrastructure/apim.tf]: done

File: infrastructure/apim.tf (hunk 6 above)
Comment (kiran-yenigala-hmcts) [reply infrastructure/apim.tf:8]: done

File: infrastructure/apim.tf (hunk 5 above)
Comment (kiran-yenigala-hmcts) [reply infrastructure/apim.tf:4]: done

File: infrastructure/apim.tf (hunk 7 above)
Comment (kiran-yenigala-hmcts) [reply infrastructure/apim.tf:41]: done

File: infrastructure/apim.tf (hunk 8 above)
Comment (kiran-yenigala-hmcts) [reply infrastructure/apim.tf:49]: will keep this here for now

File: infrastructure/apim.tf [hunk 9]
```diff
@@ -1,13 +1,54 @@
...
+  api_mgmt_name = "${var.env}" == "aat" ? "cft-api-mgmt-stg" : "cft-api-mgmt-${var.env}"
   api_mgmt_resource_group = "cft-${var.env}-network-rg"
   api_mgmt_product_name = "${var.product}-${var.component}"
   api_mgmt_api_name = "${var.product}-${var.component}-api"
-  api_base_path = var.product
+  api_base_path = "${var.product}"
   url_civil_sdt_gateway = "http://civil-sdt-gateway-${var.env}.service.core-compute-${var.env}.internal"
   url_swagger = "https://raw.githubusercontent.com/hmcts/civil-sdt-gateway/demo/src/main/resources/test.wsdl"
+
+
```
Comment (paulridout) [inline infrastructure/apim.tf:12]: Cosmetic - replace double blank line with a single blank line.

File: infrastructure/apim.tf [hunk 10]
```diff
@@ -1,13 +1,54 @@
...
 }
 
+provider "azurerm" {
+  subscription_id            = local.key_vault_subscription_id["${var.env}"].subscription
+  skip_provider_registration = "true"
+  features {}
+  alias = "key_vault_civil"
+}
+
+
```
Comment (paulridout) [inline infrastructure/apim.tf:51]: Cosmetic - replace double blank line with a single blank line.

File: infrastructure/apim.tf [hunk 11]
```diff
@@ -51,3 +92,50 @@ module "api-mgmt-api-policy" {
...
+data "azurerm_api_management" "cft-api-mgmt" {
+  name                = local.api_mgmt_name
+  resource_group_name = local.api_mgmt_resource_group
+}
+
+data "azurerm_key_vault" "key_vault" {
+  name                = local.key_vault_name
+  resource_group_name = local.key_vault_resource_group
+  provider             = azurerm.key_vault_civil
```
Comment (paulridout) [inline infrastructure/apim.tf:104]: Cosmetic - One space too many between provider and equals sign.  Terraform style guide indicates that equals signs should be aligned when there are multiple arguments - see https://developer.hashicorp.com/terraform/language/syntax/style.

**Review by paulridout — APPROVED** [review]
Changes look OK.

**Review by hemantt — APPROVED** [review]

**Pull request merged**